tests/components/airthings_ble/test_sensor.py

```diff
@@ -0,0 +1,169 @@
+"""Test the Airthings Wave sensor."""
+import logging
+
+from homeassistant.components.airthings_ble.const import DOMAIN
+from homeassistant.components.airthings_ble.sensor import async_migrate
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_registry import async_entries_for_device
+
+from tests.components.airthings_ble import (
+    CO2_V1,
+    CO2_V2,
+    HUMIDITY_V2,
+    TEMPERATURE_V1,
+    VOC_V1,
+    VOC_V2,
+    VOC_V3,
+    WAVE_DEVICE_INFO,
+    WAVE_SERVICE_INFO,
+    create_device,
+    create_entry,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def test_migration_from_v1_to_v3_unique_id(hass: HomeAssistant):
+    """User has a v1 unique id, we should migrate it to v3."""
+    entry = create_entry(hass)
+    device = create_device(hass, entry)
+
+    assert entry is not None
+    assert device is not None
+
+    entity_registry = hass.helpers.entity_registry.async_get(hass)
+
+    await hass.async_block_till_done()
+
+    entity_registry.async_get_or_create(
+        domain=DOMAIN,
+        platform="sensor",
+        unique_id=TEMPERATURE_V1.unique_id,
+        config_entry=entry,
+        device_id=device.id,
+    )
+
+    # Migrate the entities
+    async_migrate(hass, WAVE_SERVICE_INFO.address, "temperature")
+
+    entities = async_entries_for_device(
+        entity_registry,
+        device_id=device.id,
+    )
```

Comment:
This is great 👍 , but we need it to be an end-to-end test and which does not interact with any integration details

https://developers.home-assistant.io/docs/development_testing?_highlight=wri#writing-tests-for-integrations
> Make sure to not interact with any integration details in tests of integrations. Following this pattern will make the tests more robust for changes in the integration.

You can accomplish this by setting up the integration once the mocked entities are in the registry, after its finished setup, check the registry again to make sure the migration did what you expected.

Reply:
I believe there is room for further improvement in the tests, but they currently undergo end-to-end testing (as far as I know).